### PR TITLE
fix(frontend): AutoPilot notification follow-ups — branding, UX, persistence, and cross-tab sync

### PR DESCRIPTION
**Pull request by kcze**
AutoPilot (copilot) notifications had several follow-up issues after initial implementation: old "Otto" branding, UX quirks, a service-worker crash, notification state that didn't persist or sync across tabs, a broken notification sound, and noisy Sentry alerts from SSR.

### Changes 🏗️

- **Rename "Otto" → "AutoPilot"** in all notification surfaces: browser notifications, document title badge, permission dialog copy, and notification banner copy
- **Agent Activity icon**: changed from `Bell` to `Pulse` (Phosphor) in the navbar dropdown
- **Centered dialog buttons**: the "Stay in the loop" permission dialog buttons are now centered instead of right-aligned
- **Service worker notification fix**: wrapped `new Notification()` in try-catch so it degrades gracefully in service worker / PWA contexts instead of throwing `TypeError: Illegal constructor`
- **Persist notification state**: `completedSessionIDs` is now stored in localStorage (`copilot-completed-sessions`) so it survives page refreshes and new tabs
- **Cross-tab sync**: a `storage` event listener keeps `completedSessionIDs` and `document.title` in sync across all open tabs — clearing a notification in one tab clears it everywhere
- **Fix notification sound**: corrected the sound file path from `/sounds/notification.mp3` to `/notification.mp3` and added a `.gitignore` exception (root `.gitignore` has a blanket `*.mp3` ignore rule from legacy AutoGPT agent days)
- **Fix SSR Sentry noise**: guarded the Copilot Zustand store initialization with a client-side check so `storage.get()` is never called during SSR, eliminating spurious Sentry alerts (BUILDER-7CB, 7CC, 7C7) while keeping the Sentry reporting in `local-storage.ts` intact for genuinely unexpected SSR access

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] Verify "AutoPilot" appears (not "Otto") in browser notification, document title, permission dialog, and banner
  - [x] Verify Pulse icon in navbar Agent Activity dropdown
  - [x] Verify "Stay in the loop" dialog buttons are centered
  - [x] Open two tabs on copilot → trigger completion → both tabs show badge/checkmark
  - [x] Click completed session in tab 1 → badge clears in both tabs
  - [x] Refresh a tab → completed session state is preserved
  - [x] Verify notification sound plays on completion
  - [x] Verify no Sentry alerts from SSR localStorage access

### CODE REVIEW

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts [hunk 1]
```diff
@@ -115,4 +139,27 @@ export function useCopilotNotifications(activeSessionID: string | null) {
...
+  useEffect(() => {
+    function handleStorage(e: StorageEvent) {
+      if (e.key !== Key.COPILOT_COMPLETED_SESSIONS) return;
+      let next: Set<string>;
+      try {
+        next = e.newValue
+          ? new Set<string>(JSON.parse(e.newValue))
+          : new Set<string>();
+      } catch {
+        next = new Set<string>();
```
[the remark this answers is not in the document]
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: Fixed in f5df187 — added `Array.isArray` + `typeof === 'string'` filter in the storage event handler, mirroring the validation in `loadCompletedSessions()`.

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts [hunk 2]
```diff
@@ -115,4 +139,27 @@ export function useCopilotNotifications(activeSessionID: string | null) {
...
+      }
+      useCopilotUIStore.setState({ completedSessionIDs: next });
+      const count = next.size;
+      document.title =
+        count > 0
+          ? `(${count}) AutoPilot is ready - ${ORIGINAL_TITLE}`
+          : ORIGINAL_TITLE;
```
[the remark this answers is not in the document]
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: Considered this, but a merge (union) approach would break removal propagation — when a user clears a completed session in Tab 1, Tab 2 needs to adopt the reduced set, not union it back.

localStorage is the shared source of truth here. The \last-write-wins\ race requires two tabs to process different WS completions and persist within the same JS event loop tick, which is practically impossible since persistence is synchronous within the Zustand setter. Added a comment clarifying the design choice.

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts [hunk 3]
```diff
@@ -115,4 +139,27 @@ export function useCopilotNotifications(activeSessionID: string | null) {
...
+
+  // Sync completedSessionIDs across tabs via localStorage storage events
+  useEffect(() => {
+    function handleStorage(e: StorageEvent) {
+      if (e.key !== Key.COPILOT_COMPLETED_SESSIONS) return;
+      let next: Set<string>;
+      try {
+        next = e.newValue
+          ? new Set<string>(JSON.parse(e.newValue))
+          : new Set<string>();
```
[the remark this answers is not in the document]
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: This is intentional behavior, not a bug. The ChatSidebar `useEffect` clears the completed indicator when you navigate to (or are already viewing) a completed session — the user is already looking at the response, so the notification badge should clear.

If Tab 2 is viewing the same session when completion arrives, the indicator correctly appears briefly then clears because the user is already on that session. The localStorage write from `clearCompletedSession` then propagates to Tab 1, which is also correct — the session was viewed.

File: autogpt_platform/frontend/src/app/(platform)/copilot/store.ts [hunk 4]
```diff
@@ -6,6 +6,28 @@ export interface DeleteTarget {
...
+const isClient = typeof window !== "undefined";
+
+function loadCompletedSessions(): Set<string> {
+  if (!isClient) return new Set();
+  const raw = storage.get(Key.COPILOT_COMPLETED_SESSIONS);
+  if (!raw) return new Set();
+  try {
+    const parsed = JSON.parse(raw);
+    return Array.isArray(parsed) ? new Set(parsed) : new Set();
+  } catch {
```
[the remark this answers is not in the document]
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/store.ts]: Fixed in 761608a — added `typeof v === 'string'` filter, mirroring the storage event handler validation.

File: autogpt_platform/frontend/src/app/(platform)/copilot/store.ts [hunk 5]
```diff
@@ -6,6 +6,28 @@ export interface DeleteTarget {
...
+  }
+}
+
+function persistCompletedSessions(ids: Set<string>) {
+  if (ids.size === 0) {
+    storage.clean(Key.COPILOT_COMPLETED_SESSIONS);
+  } else {
+    storage.set(Key.COPILOT_COMPLETED_SESSIONS, JSON.stringify([...ids]));
+  }
+}
```
[the remark this answers is not in the document]
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/store.ts:23]: Fixed in 761608a — added `isClient` guard and try-catch to `persistCompletedSessions`.

**Review by ntindle — APPROVED** [review]
dont forget to test it

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts [hunk 6]
```diff
@@ -49,7 +83,7 @@ export function useCopilotNotifications(activeSessionID: string | null) {
       // Always update UI state (checkmark + title) regardless of notification setting
       state.addCompletedSession(sessionID);
       const count = useCopilotUIStore.getState().completedSessionIDs.size;
-      document.title = `(${count}) Otto is ready - ${ORIGINAL_TITLE}`;
+      document.title = `(${count}) AutoPilot is ready - ${ORIGINAL_TITLE}`;
```
Comment (Abhi1992002) [inline autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: 🟡 **Nice to Have**: The document title format string `(${count}) AutoPilot is ready - ${ORIGINAL_TITLE}` is repeated in 4 places across this PR:

1. `ChatSidebar.tsx` line 128 (hardcodes `"AutoGPT"` instead of using `ORIGINAL_TITLE`)
2. `useCopilotNotifications.ts` line 86 (WebSocket handler)
3. `useCopilotNotifications.ts` line 51 (init useEffect)
4. `useCopilotNotifications.ts` line 169 (storage event handler)

Consider extracting a shared helper like:

```ts
function formatNotificationTitle(count: number): string {
  return count > 0 ? `(${count}) AutoPilot is ready - ${ORIGINAL_TITLE}` : ORIGINAL_TITLE;
}
```

This would also fix the inconsistency where ChatSidebar hardcodes `"AutoGPT"` while the other locations use the `ORIGINAL_TITLE` constant.

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts [hunk 7]
```diff
@@ -115,4 +142,40 @@ export function useCopilotNotifications(activeSessionID: string | null) {
...
+        if (!e.newValue) {
+          next = new Set<string>();
+        } else {
+          const parsed: unknown = JSON.parse(e.newValue);
+          next = Array.isArray(parsed)
+            ? new Set<string>(parsed.filter((v) => typeof v === "string"))
+            : new Set<string>();
+        }
+      } catch {
+        next = new Set<string>();
```
Comment (Abhi1992002) [inline autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: 🔵 **Nit**: The JSON parsing + validation logic (JSON.parse → Array.isArray → string filter) is duplicated between `loadCompletedSessions()` in `store.ts` and the `handleStorage` listener here. Could extract a shared `parseSessionIDs(raw: string | null): Set<string>` helper to keep both in sync.

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts (hunk 6 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: Fixed in c1407f38c — extracted `formatNotificationTitle(count)` into `copilot/helpers.ts` and replaced all 4 call sites (including the hardcoded `"AutoGPT"` in ChatSidebar). Also moved `ORIGINAL_TITLE` to the shared helpers so `store.ts`'s `clearCopilotLocalData` uses it too.

File: autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts (hunk 7 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/useCopilotNotifications.ts]: Fixed in c1407f38c — extracted `parseSessionIDs(raw)` into `copilot/helpers.ts` and updated both `store.ts` (init) and `useCopilotNotifications.ts` (storage event handler) to use it.

**Comment by majdyz** [comment]
## 🤖 PR Review — #12428

**AutoPilot notification follow-ups — Branding, UX, persistence, cross-tab sync improvements.**

CI: 0 failures | Diff: ~364 lines

Review in progress — will post inline findings if any.

**Comment by majdyz** [comment]
## 🤖 PR Review — #12428 (Notification Follow-ups)

**Verdict: ✅ APPROVED**

UX improvements for AutoPilot notifications. Sound notification, cross-tab sync via localStorage, branding updates. Well-scoped changes across notification components.

**Findings:** No blockers.

All CI checks green.

File: autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationBanner/NotificationBanner.tsx [hunk 8]
```diff
@@ -56,8 +56,8 @@ export function NotificationBanner() {
     <div className="flex items-center gap-3 border-b border-amber-200 bg-amber-50 px-4 py-2.5">
       <BellRinging className="h-5 w-5 shrink-0 text-amber-600" weight="fill" />
       <Text variant="body" className="flex-1 text-sm text-amber-800">
-        Enable browser notifications to know when Otto finishes working, even
-        when you switch tabs.
+        Enable browser notifications to know when AutoPilot finishes working,
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationBanner/NotificationBanner.tsx:59]: 🤖 🔵 **Nit**: Good catch renaming "Otto" to "AutoPilot" for brand consistency across all notification surfaces.

File: autogpt_platform/frontend/src/app/(platform)/copilot/store.ts [hunk 9]
```diff
@@ -1,11 +1,27 @@
 import { Key, storage } from "@/services/storage/local-storage";
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/store.ts:1]: 🤖 🟡 **Nice to Have**: Cross-tab notification sync via localStorage is solid. Consider debouncing the storage event listener to avoid rapid state updates when multiple tabs write simultaneously.

File: autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationDialog/NotificationDialog.tsx [hunk 10]
```diff
@@ -77,11 +77,12 @@ export function NotificationDialog() {
...
           <Text variant="body" className="text-center text-neutral-600">
-            Otto can notify you when a response is ready, even if you switch
-            tabs or close this page. Enable notifications so you never miss one.
+            AutoPilot can notify you when a response is ready, even if you
+            switch tabs or close this page. Enable notifications so you never
+            miss one.
           </Text>
         </div>
-        <Dialog.Footer>
+        <Dialog.Footer className="justify-center">
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationDialog/NotificationDialog.tsx:85]: 🤖 🟠 **Should Fix**: `Dialog.Footer` (via `BaseFooter`) renders `justify-end` as a hardcoded base class, and this PR passes `justify-center` via `className`. Since `BaseFooter` uses raw string concatenation (`` `flex justify-end gap-4 pt-6 ${className}` ``) instead of `cn()` / `tailwind-merge`, both `justify-end` and `justify-center` end up in the class list. Which one wins depends on the CSS cascade order in the generated Tailwind stylesheet, **not** the order in the HTML class attribute — so the result is unpredictable and may silently break across Tailwind version bumps or config changes.

Two options to fix:
1. Update `BaseFooter` to use `cn()` instead of raw string concatenation (preferred — fixes the root cause for all callers).
2. Or use inline `style={{ justifyContent: "center" }}` which will always win over class-based styling.

File: autogpt_platform/frontend/src/app/(platform)/copilot/store.ts [hunk 11]
```diff
@@ -85,11 +109,12 @@ export const useCopilotUIStore = create<CopilotUIState>((set) => ({
...
     storage.clean(Key.COPILOT_NOTIFICATION_BANNER_DISMISSED);
     storage.clean(Key.COPILOT_NOTIFICATION_DIALOG_DISMISSED);
+    storage.clean(Key.COPILOT_COMPLETED_SESSIONS);
     set({
       completedSessionIDs: new Set<string>(),
       isNotificationsEnabled: false,
       isSoundEnabled: true,
     });
-    document.title = "AutoGPT";
+    document.title = ORIGINAL_TITLE;
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/store.ts]: 🤖 🟡 **Nice to Have**: `clearCopilotLocalData` accesses `document.title` without an `isClient` guard. Every other SSR-sensitive path in this file is now protected (`completedSessionIDs` init, `isNotificationsEnabled`, `isSoundEnabled`), but this one was missed. Low risk since it is only called from client components, but adding `if (isClient)` before `document.title = ORIGINAL_TITLE` would complete the SSR hardening story.

File: autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts [hunk 12]
```diff
@@ -1,5 +1,33 @@
...
+  if (!raw) return new Set();
+  try {
+    const parsed: unknown = JSON.parse(raw);
+    return Array.isArray(parsed)
+      ? new Set<string>(parsed.filter((v) => typeof v === "string"))
+      : new Set();
+  } catch {
+    return new Set();
+  }
+}
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts:29]: 🤖 🟠 **Should Fix**: `formatNotificationTitle` and `parseSessionIDs` are pure functions with clear edge cases (negative/NaN count, empty string, malformed JSON, non-array JSON, array with non-string elements) but have no unit tests. These are the kind of functions that are trivial to test and will catch regressions if someone refactors them later. Please add a colocated `helpers.test.ts` covering the main paths and edges.

File: autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts [hunk 13]
```diff
@@ -1,5 +1,33 @@
 import type { UIMessage } from "ai";
 
+export const ORIGINAL_TITLE = "AutoGPT";
+
+/**
+ * Build the document title showing how many sessions are ready.
+ * Returns the base title when count is 0.
+ */
+export function formatNotificationTitle(count: number): string {
+  return count > 0
```
Comment (majdyz) [inline autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts:10]: 🤖 🟡 **Nice to Have**: `formatNotificationTitle` only guards against `count <= 0` by checking `count > 0`, so negative values (which could theoretically appear from `completedSessionIDs.size - 1` when the set is empty) correctly fall through to the base title. That said, the callers in `ChatSidebar` compute `remaining = completedSessionIDs.size - 1` — if `clearCompletedSession` is called for an ID that was already removed (e.g. double-click or cross-tab race), `remaining` could go negative. Consider using `Math.max(0, remaining)` at the call site for safety.

**Comment by majdyz** [comment]
When the feedback is addressed, please re-request review so I can take another look.

File: autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationDialog/NotificationDialog.tsx (hunk 10 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/components/NotificationDialog/NotificationDialog.tsx:85]: Fixed in 34e157f — updated `BaseFooter` to use `cn()` instead of raw string concatenation, so Tailwind class overrides like `justify-center` now properly merge.

File: autogpt_platform/frontend/src/app/(platform)/copilot/store.ts (hunk 11 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/store.ts]: Fixed in 34e157f — added `isClient` guard around `document.title = ORIGINAL_TITLE`.

File: autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts (hunk 12 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts:29]: Fixed in 34e157f — added `helpers.test.ts` with 15 tests covering both functions: zero/positive/negative/NaN counts for `formatNotificationTitle`, and null/undefined/empty/valid/non-string/non-array/malformed/dedup cases for `parseSessionIDs`.

File: autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts (hunk 13 above)
Comment (kcze) [reply autogpt_platform/frontend/src/app/(platform)/copilot/helpers.ts:10]: Fixed in 34e157f — added `Math.max(0, completedSessionIDs.size - 1)` at the call site in ChatSidebar.

File: .gitignore [hunk 14]
```diff
@@ -16,6 +16,7 @@ log-ingestion.txt
 /logs
 *.log
 *.mp3
+!autogpt_platform/frontend/public/notification.mp3
```
Comment (majdyz) [inline .gitignore:19]: what is this?

Comment (kcze) [reply .gitignore:19]: AutoPilot notification sound that was ignored in the original PR 🥲
